R/derive_param_wbc_abs.R

```diff
@@ -0,0 +1,162 @@
+#' "P01",    "LYMLE",   70,   "Lymphocytes (fraction of 1)",    "CYCLE 2 DAY 1",
+#' "P01",    "ALB",     36,   "Albumin (g/dL)",                 "CYCLE 2 DAY 1",
+#' "P02",    "WBC",     33,   "Leukocyte Count (10^9/L)",       "CYCLE 1 DAY 1",
+#' "P02",    "LYMPH",   29,   "Leukocyte Count (10^9/L)",       "CYCLE 1 DAY 1",
+#' "P02",    "LYMLE",   87,   "Lymphocytes (fraction of 1)",    "CYCLE 1 DAY 1",
+#' "P03",    "LYMLE",   89,   "Lymphocytes (fraction of 1)",    "CYCLE 1 DAY 1"
+#' )
+#'
+#' derive_param_wbc_abs (dataset = test_lb,
+#'                       by_vars = vars(USUBJID, VISIT),
+#'                       set_values_to = vars(
+#'                         PARAMCD = "LYMPH",
+#'                         PARAM = "Lymphocytes Abs (10^9/L)",
+#'                         DTYPE = "CALCULATION"
+#'                       ),
+#'                       get_unit_expr = extract_unit(PARAM),
+#'                       wbc_code = "WBC",
+#'                       diff_code = "LYMLE")
+
+derive_param_wbc_abs <- function(dataset,
+                                 by_vars,
+                                 set_values_to,
+                                 get_unit_expr,
+                                 wbc_unit = "10^9/L",
+                                 diff_unit = "fraction of 1",
+                                 wbc_code = "WBC",
+                                 diff_code) {
+  assert_vars(by_vars)
+  assert_data_frame(dataset, required_vars = vars(!!!by_vars, PARAMCD, AVAL))
+  assert_character_scalar(wbc_code)
+  assert_character_scalar(diff_code)
+  get_unit_expr <- assert_expr(enquo(get_unit_expr))
+  assert_character_scalar(wbc_unit)
+  assert_character_scalar(diff_unit)
+  by_vars_str <- vars2chr(by_vars)
+
+  # Check unit of input parameters
+  assert_unit(dataset, wbc_code, required_unit = wbc_unit, get_unit_expr = !!get_unit_expr)
+  assert_unit(dataset, diff_code, required_unit = diff_unit, get_unit_expr = !!get_unit_expr)
+
+  # select observations and variables required for new observations.
+  dataset_temp <- dataset %>%
+    filter(
+      PARAMCD == !!wbc_code |
+      PARAMCD == !!diff_code |
+      PARAMCD == !!quo_get_expr(set_values_to$PARAMCD)
+    ) %>%
+    select(by_vars_str, PARAMCD, AVAL)
+
+  # only keep records where absolute value do not already exist.
+  dataset_abs <- dataset_temp %>%
+    filter(PARAMCD == !!quo_get_expr(set_values_to$PARAMCD)) %>%
+    mutate(temp_flag = "Y") %>%
+    select(by_vars_str, temp_flag)
+
+  dataset_temp <- dataset_temp %>%
+    left_join(dataset_abs, by = by_vars_str) %>%
+    filter(is.na(temp_flag))
+
+  # create new parameters.
+  temp_set_values_to <- quo_c(set_values_to, vars(temp_var = "new"))
+
+  dataset_new <- dataset_temp %>%
+    derive_derived_param(
+      parameters = c(wbc_code, diff_code),
+      by_vars = by_vars,
+      analysis_value = ((!!sym(paste0("AVAL.", wbc_code)) *
+                           !!sym(paste0("AVAL.", diff_code))) / 100),
```

Comment:
I removed the `diff_unit` argument, and added a `diff_type` argument that takes in either `"fraction"` or `"percent"`. If `"fraction"` the function doesn't divide by 100, if "percent" it divides by 100. However, I kept the `wbc_unit` since it defaults to "10^9/L" which I think is the common unit? But if you think it's better to remove the unit checking I can do that too.